#! /bin/sh /usr/share/dpatch/dpatch-run
## 01_secure_sessions.dpatch by  <svenud@ozemail.com.au>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP:   * secure the session files, and use file time to expire them 
## DP:       Arbitrary code execution in session files (CVE-2007-0669) (Closes #410256)

@DPATCH@
diff -urNad un~/lib/TWiki/Client.pm un/lib/TWiki/Client.pm
--- un~/lib/TWiki/Client.pm	2006-10-25 10:16:05.000000000 +1000
+++ un/lib/TWiki/Client.pm	2007-02-10 23:55:15.000000000 +1100
@@ -338,42 +338,28 @@
 =cut
 
 sub expireDeadSessions {
-	my $time = time() || 0;
+    my $time = time() || 0;
     my $exp = $TWiki::cfg{Sessions}{ExpireAfter} || 36000; # 10 hours
     $exp = -$exp if $exp < 0;
 
-	opendir(D, $TWiki::cfg{Sessions}{Dir}) || return;
-	foreach my $file ( grep { /cgisess_[0-9a-f]{32}/ } readdir(D) ) {
+    my $tmpFile = $TWiki::cfg{Sessions}{Dir};
+    opendir(D, $tmpFile) || return;
+    foreach my $file ( grep { /^(passthru|cgisess)_[0-9a-f]{32}/ } readdir(D) ) {
         $file = TWiki::Sandbox::untaintUnchecked(
-            $TWiki::cfg{Sessions}{Dir}.'/'.$file );
-		my @stat = stat( $file );
-        # Kill small old files. They can't be valid sessions.
-		# Ignore tiny new files. They can't be complete sessions.
-        if( defined($stat[7]) && $stat[7] <= 50 ) {
-            my $lat = $stat[8] || $stat[9] || $stat[10] || 0;
-            unlink $file if( $time - $lat >= $exp );
-            next;
-		}
-
-		open(F, $file) || next;
-		my $session = <F>;
-		close F;
-
-        # SMELL: security hazard?
-        $session = TWiki::Sandbox::untaintUnchecked( $session );
-
-        my $D;
-		eval $session;
-		next if ( $@ );
-        # The session is expired if it is empty, hasn't been accessed in ages
-        # or has exceeded its registered expiry time.
-        if( !$D || $time >= $D->{_SESSION_ATIME} + $exp ||
-              $D->{_SESSION_ETIME} && $time >= $D->{_SESSION_ETIME} ) {
-            unlink( $file );
-            next;
-        }
-	}
-	closedir D;
+            $tmpFile.'/'.$file );
+        my @stat = stat( $file );
+        # CGI::Session updates the session file each time a browser views a
+        # topic setting the access and expiry time as values in the file. This 
+        # also sets the mtime (modification time) for the file which is all we need.
+        # We know that the expiry time is mtime + $TWiki::cfg{Sessions}{ExpireAfter}
+        # so we do not need to waste execution time opening and reading the file.
+        # We just check the mtime. mtime is confirmed set in both Windows and Linux
+        # As a fallback we also check ctime. Files are deleted when they expire.
+        my $lat = $stat[9] || $stat[10] || 0;
+        unlink $file if ( $time - $lat >= $exp );
+        next;
+    }
+    closedir D;
 }
 
 =pod
